File: prices/producers/node/produce.js
const fs = require("fs");
const { Kafka } = require("kafkajs");
require('dotenv').config()


const { KAFKA_BOOTSTRAP_SERVER, KAFKA_SSL_CAFILE } = process.env;

const kafka = new Kafka({
    clientId: "nodejs-price-producer",
    brokers: [KAFKA_BOOTSTRAP_SERVER],
    ssl: {
        ca: [fs.readFileSync(KAFKA_SSL_CAFILE, "utf-8")],
    }
});


produce();


async function produce() {
    const producer = kafka.producer();

    await producer.connect();

    send();

    async function send() {
        // Send an int value as bytes (32bit Big Endian integer)
        const price = getRandom(1000, 2000);
        const value = intToBytes(price);

        await producer.send({
            topic: 'prices',
            messages: [{ value }]
        });

        console.log("Sent", price);

        setTimeout(send, 2000);
    }
}


function intToBytes(price) {
    const value = Buffer.allocUnsafe(4);
    value.writeInt32BE(price);
    return value;
}

function getRandom(min, max) {
    return Math.floor(Math.random() * (max - min) + min);
}

File: prices/producers/python/produce.py
import os
from time import sleep
from random import randint
from kafka import KafkaProducer
from dotenv import load_dotenv


# Read environment variables from .env
load_dotenv()


def int_serializer(x):
    return x.to_bytes(4, 'big')


producer = KafkaProducer(
    bootstrap_servers=os.environ["KAFKA_BOOTSTRAP_SERVER"],
    value_serializer=int_serializer,
    security_protocol="SSL",
    ssl_cafile=os.environ["KAFKA_SSL_CAFILE"],
)


while True:
    price = randint(1, 100)
    producer.send('prices', price)
    print("Sent", price)
    sleep(1)
